pkg/types/types.go

```diff
@@ -76,3 +78,36 @@ type MercuryProvider interface {
 	OnchainConfigCodec() mercury.OnchainConfigCodec
 	ContractTransmitter() mercury.Transmitter
 }
+
+type FunctionsProvider interface {
+	PluginProvider
+	LogPollerWrapper() LogPollerWrapper
+	NodeSignerKey() *ecdsa.PrivateKey
```

Comment:
How is this key used? We have been trying to avoid passing keys across this API, and I'm surprised to see a key owned by the relayer :thinking: - the normal keystore lives in core and we just provide a signing API.

Reply:
Functions needs a signer key for Gateway communication. It has a node address in the config and then finds the corresponding key in keystore.

Can this exist in the delegate? https://github.com/smartcontractkit/chainlink/blob/develop/core/services/ocr2/delegate.go#L80

It's evm-specific so I assumed it needs to go through relay.

Reply:
We pass a keystore handle to the relayers when we construct them - we can do the same with the functions plugin.
Must you have access to the raw key? Or could you work with some kind of signing API? We have this currently for signing txes:
https://github.com/smartcontractkit/chainlink-relay/blob/main/pkg/loop/internal/relayer.go#L18-L22
```go
type Keystore interface {
	Accounts(ctx context.Context) (accounts []string, err error)
	// Sign returns data signed by account.
	// nil data can be used as a no-op to check for account existence.
	Sign(ctx context.Context, account string, data []byte) (signed []byte, err error)
}
```
And we can pack structured data in to the bytes as necessary. (e.g. starknet actually wants some big.Ints to be returned)

Reply:
If you do need the private key, we could also just pass it directly when we construct the plugin (constructor arg or env var?) rather than expand this API that sort of serves a different purpose.

Reply:
Ideally we use the interface? it feels more future proof than passing the private key around - especially for loop work.

we've had annoying issues with circular imports that were solved once we put things behind a signer interface (won't happen here cuz ecdsa is a 3rd party dependency, but custom key definitions in core imported into relayers have come to sting us a few times 🐝)

Reply:
Excellent!